src/cli/cli.cpp

```diff
@@ -4416,58 +4256,79 @@ void Interpreter::ProcessDiag(uint8_t aArgsLength, char *aArgs[])
+
+    while (left < right)
+    {
+        uint16_t middle  = (left + right) / 2;
+        int      compare = strcmp(aName, sCommands[middle].mName);
+
+        if (compare == 0)
+        {
+            rval = &sCommands[middle];
+            break;
+        }
+        else if (compare > 0)
+        {
+            left = middle + 1;
+        }
+        else
+        {
+            right = middle;
+        }
+    }
+
+    return rval;
+}
+
 void Interpreter::ProcessLine(char *aBuf, uint16_t aBufLength)
 {
-    char *  aArgs[kMaxArgs] = {nullptr};
-    char *  cmd;
-    uint8_t aArgsLength = 0;
-    size_t  i           = 0;
+    char *         aArgs[kMaxArgs] = {nullptr};
+    char *         cmdName;
+    uint8_t        aArgsLength = 0;
+    const Command *command;
 
     VerifyOrExit(aBuf != nullptr && StringLength(aBuf, aBufLength + 1) <= aBufLength, OT_NOOP);
 
     VerifyOrExit(Utils::CmdLineParser::ParseCmd(aBuf, aArgsLength, aArgs, kMaxArgs) == OT_ERROR_NONE,
-                 OutputFormat("Error: too many args (max %d)\r\n", kMaxArgs));
-    VerifyOrExit(aArgsLength >= 1, OutputFormat("Error: no given command.\r\n"));
+                 OutputLine("Error: too many args (max %d)", kMaxArgs));
+    VerifyOrExit(aArgsLength >= 1, OutputLine("Error: no given command."));
 
-    cmd = aArgs[0];
+    cmdName = aArgs[0];
 
 #if OPENTHREAD_CONFIG_DIAG_ENABLE
-    VerifyOrExit((!otDiagIsEnabled(mInstance) || (strcmp(cmd, "diag") == 0)),
-                 OutputFormat("under diagnostics mode, execute 'diag stop' before running any other commands.\r\n"));
+    VerifyOrExit((!otDiagIsEnabled(mInstance) || (strcmp(cmdName, "diag") == 0)),
+                 OutputLine("under diagnostics mode, execute 'diag stop' before running any other commands."));
 #endif
 
-    for (i = 0; i < OT_ARRAY_LENGTH(sCommands); i++)
+    command = FindCommand(cmdName);
+
+    if (command != nullptr)
     {
-        if (strcmp(cmd, sCommands[i].mName) == 0)
-        {
-            (this->*sCommands[i].mCommand)(aArgsLength - 1, &aArgs[1]);
-            break;
-        }
+        (this->*command->mCommand)(aArgsLength - 1, &aArgs[1]);
+        ExitNow();
     }
 
-    // Check user defined commands if built-in command
-    // has not been found
-    if (i == OT_ARRAY_LENGTH(sCommands))
+    // Check user defined commands if built-in command has not been found
+    for (uint8_t i = 0; i < mUserCommandsLength; i++)
     {
-        for (i = 0; i < mUserCommandsLength; i++)
+        if (strcmp(cmdName, mUserCommands[i].mName) == 0)
         {
-            if (strcmp(cmd, mUserCommands[i].mName) == 0)
-            {
-                mUserCommands[i].mCommand(aArgsLength - 1, &aArgs[1]);
-                break;
-            }
-        }
-
-        if (i == mUserCommandsLength)
-        {
-            AppendResult(OT_ERROR_INVALID_COMMAND);
+            mUserCommands[i].mCommand(aArgsLength - 1, &aArgs[1]);
+            ExitNow();
```

Comment:
Change `FindCommand` so that it could also be applied to `UserCommands`?

Reply:
We cannot easily extend `FindCommand()` to use for both  `sCommand` and `mUserCommands` arrays:
- Unlike the `sCommands` we cannot know if the `mUserCommands` list is sorted or not at compile time (it's given to us at run-time), so cannot always use binary search for it.
- Another point is that their underlying types are different (one uses `Command` struct and the other one uses `otCliCommand`), most notably their handler type `mCommand` are different:
   - `Command` uses an `Interpreter` method pointer 
   ```void (Interpreter::*mCommand)(uint8_t aArgsLength, char *aArgs[])``` 
   -  but `otCliCommand` uses a normal function pointer `void (*mCommand)(uint8_t aArgsLength, char *aArgs[])`

Reply:
Got it. Thanks.